common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java

```diff
@@ -373,6 +427,54 @@ public ManagedBuffer next() {
     }
   }
 
+  /**
+   * Dummy implementation of merged shuffle file manager. Suitable for when push-based shuffle
+   * is not enabled.
+   */
+  private static class NoOpMergedShuffleFileManager implements MergedShuffleFileManager {
+
+    @Override
+    public StreamCallbackWithID receiveBlockDataAsStream(PushBlockStream msg) {
+      throw new UnsupportedOperationException("Cannot handle shuffle block merge");
+    }
+
+    @Override
+    public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOException {
+      throw new UnsupportedOperationException("Cannot handle shuffle block merge");
+    }
+
+    @Override
+    public void registerApplication(String appId, String user) {
```

Comment:
Why we need `user` here? Is it a system user or something else?

Reply:
The `user` is the ID for the user running the Spark app. The `appId` and `user` together indicates the unique local dir paths where merged shuffle files for executors on a given node are stored.

Is your concern that this API might be too YARN specific?

Reply:
that is definitely a good question. For other external shuffle services is this enough or should it be behind another interface that different ones could specify different arguments.

Reply:
It's a bit unclear at this moment, especially on that part of what's needed in different schedulers.
Our current approach for determining the merged shuffle file directory path is the following:

1. The implementation of MergedShuffleFileManager (RPC handler for block push requests) will be initialized with a relative directory path pattern, which is relative to the list of executor local dirs (a common concept across all schedulers).
2. The actual path for storing the merged shuffle files for a given application on a given host is then decided based on the local dirs and the materialization of the relative path pattern with the appId and user ID.

The assumption is that once we know the local dirs for a given app, the remaining portion of the directory path to the merged shuffle files will be mostly the same across different applications except the app Id and the user Id portion in the path.

Reply:
I'd prefer to resolve this comment for now, and revisit when it's more clear what the requirement for  a potentially more generic API would be.